pkg/controller/notification/mailgun_notification_delivery_service.go

```diff
@@ -2,17 +2,82 @@ package notification
 
 import (
 	"context"
+	"errors"
+	"fmt"
+	"github.com/codeready-toolchain/host-operator/pkg/templates/notificationtemplates"
+	"github.com/mailgun/mailgun-go/v4"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
 )
 
 type MailgunNotificationDeliveryService struct {
+	base        BaseNotificationDeliveryService
+	Mailgun     mailgun.Mailgun
+	SenderEmail string
 }
 
-// NewMailgunNotificationDeliveryService
-func NewMailgunNotificationDeliveryService() NotificationDeliveryService {
-	return &MailgunNotificationDeliveryService{}
+type MailgunConfiguration interface {
+	GetMailgunDomain() string
+	GetMailgunAPIKey() string
+	GetMailgunSenderEmail() string
 }
 
-func (s *MailgunNotificationDeliveryService) Send(ctx context.Context, notificationCtx NotificationContext, templateName string) error {
-	// TODO implement
-	return nil
+type MailgunOption interface {
+	// ApplyToMailgun applies this configuration to the given mailgun instance.
+	ApplyToMailgun(mailgun.Mailgun)
+}
+
+// NewMailgunNotificationDeliveryService creates a delivery service that uses the Mailgun API to deliver email notifications
+func NewMailgunNotificationDeliveryService(client client.Client, config MailgunConfiguration,
+	opts ...MailgunOption) NotificationDeliveryService {
+
+	mg := mailgun.NewMailgun(config.GetMailgunDomain(), config.GetMailgunAPIKey())
+
+	for _, opt := range opts {
+		opt.ApplyToMailgun(mg)
+	}
+
+	svc := &MailgunNotificationDeliveryService{
+		base:        BaseNotificationDeliveryService{},
+		Mailgun:     mg,
+		SenderEmail: config.GetMailgunSenderEmail(),
+	}
+
+	return svc
+}
+
+func (s *MailgunNotificationDeliveryService) Send(ctx context.Context, notificationCtx *NotificationContext, templateName string) error {
+
+	template, found, err := notificationtemplates.GetNotificationTemplate(templateName)
+	if err != nil {
+		return err
+	}
+
+	if !found {
+		return errors.New(fmt.Sprintf("notification template [%s] not found", templateName))
+	}
+
+	sender := s.SenderEmail
+
+	subject, err := s.base.GenerateContent(ctx, notificationCtx, template.Subject)
+	if err != nil {
+		return err
+	}
+
+	body, err := s.base.GenerateContent(ctx, notificationCtx, template.Content)
+	if err != nil {
+		return err
+	}
+
+	recipient := notificationCtx.FullEmailAddress()
+
+	// The message object allows you to add attachments and Bcc recipients
+	message := s.Mailgun.NewMessage(sender, subject, body, recipient)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+
+	// Send the message with a 10 second timeout
+	_, _, err = s.Mailgun.Send(ctx, message)
```

Comment:
I'm not quite sure what to do with these.  We could log them at this point here, but I don't know what the value of doing that would be.  We can't really return them because they're only meaningful within the mailgun delivery service implementation.  I would expect that if the response was something other than a successful send, then the Send() function would return an error anyway. @alexeykazakov any thoughts?

Reply:
We need to check what errors `Send()` can return. If is it enough to relay on the `err` here or we really need to analyze the response too?
But anyway, in case of the error I think it could be helpful if we log the response body and the id (is it the request ID?) here. No need to return them IMO.

Reply:
Sounds good. if we're going to log this error here should we also log the others above? 
```
template, found, err := s.base.TemplateLoader.GetNotificationTemplate(templateName)
subject, err := s.base.GenerateContent(notificationCtx, template.Subject)
body, err := s.base.GenerateContent(notificationCtx, template.Content)
```

Reply:
what we can do actually is to wrap this error with out own custom own with additional information and return so it will be properly logged and added to the status

Reply:
I've introduced a new custom error type and now return that if there's an error